pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go

```diff
@@ -0,0 +1,662 @@
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	// ManagedResourceName is the name of the ManagedResource containing the resource specifications.
+	ManagedResourceName = "vpn-seed-server"
+
+	DeploymentName = v1beta1constants.DeploymentNameVPNSeedServer
+)
+
+// VPNSeedServer contains functions for a vpn-seed-server deployer.
+type VPNSeedServer interface {
+	component.DeployWaiter
+
+	// SetSecrets sets the secrets.
+	SetSecrets(Secrets)
+}
+
+// Secrets is collection of secrets for the vpn-seed-server.
+type Secrets struct {
+	// TLSAuth is a secret containing the TLSAuth certificate.
+	TLSAuth component.Secret
+	// Server is a secret containing the server certificate and key.
+	Server component.Secret
+	// DH is a secret containing the DH key.
+	DH component.Secret
+}
+
+// New creates a new instance of DeployWaiter for the vpn-seed-server.
+func New(
+	client client.Client,
+	namespace string,
+	imageAPIServerProxy string,
+	imageVPNSeedServer string,
+	kubeAPIServerHost *string,
+	serviceNetwork string,
+	podNetwork string,
+	nodeNetwork *string,
+	replicas int32,
+) VPNSeedServer {
+	return &vpnSeedServer{
+		client:              client,
+		namespace:           namespace,
+		imageAPIServerProxy: imageAPIServerProxy,
+		imageVPNSeedServer:  imageVPNSeedServer,
+		kubeAPIServerHost:   kubeAPIServerHost,
+		serviceNetwork:      serviceNetwork,
+		podNetwork:          podNetwork,
+		nodeNetwork:         nodeNetwork,
+		replicas:            replicas,
+	}
+}
+
+type vpnSeedServer struct {
+	client              client.Client
+	namespace           string
+	imageAPIServerProxy string
+	imageVPNSeedServer  string
+	kubeAPIServerHost   *string
+	serviceNetwork      string
+	podNetwork          string
+	nodeNetwork         *string
+	replicas            int32
+
+	secrets Secrets
+}
+
+func (v *vpnSeedServer) Deploy(ctx context.Context) error {
+
+	if v.secrets.TLSAuth.Name == "" || v.secrets.TLSAuth.Checksum == "" {
+		return fmt.Errorf("missing TLSAuth secret information")
+	}
+	if v.secrets.DH.Name == "" || v.secrets.DH.Checksum == "" {
+		return fmt.Errorf("missing DH secret information")
+	}
+	if v.secrets.Server.Name == "" || v.secrets.Server.Checksum == "" {
+		return fmt.Errorf("missing server secret information")
+	}
+
+	var (
+		configMap       = v.emptyConfigMap()
+		serverSecret    = v.emptySecret(DeploymentName)
+		tlsAuthSecret   = v.emptySecret(fmt.Sprintf("%s-tlsauth", DeploymentName))
+		dhSecret        = v.emptySecret(fmt.Sprintf("%s-dh", DeploymentName))
+		service         = v.emptyService()
+		deployment      = v.emptyDeployment()
+		networkPolicy   = v.emptyNetworkPolicy()
+		gateway         = v.emptyGateway()
+		virtualService  = v.emptyVirtualService()
+		destinationRule = v.emptyDestinationRule()
```

Comment:
what about a VPA?

Reply:
Added vertical pod autoscaling of the deployment with the latest commit.